index.js

```diff
@@ -112,6 +141,56 @@ class Conf {
 		}
 	}
 
+	_migrate(migrations, versionToMigrate) {
+		const MIGRATION_KEY = '__conf-migrated-version__';
+
+		const previousMigratedVersion = this.get(MIGRATION_KEY, '0.0.0');
+
+		if (semver.eq(previousMigratedVersion, versionToMigrate)) {
+			return;
+		}
+
+		const newerVersions = Object.keys(migrations)
+			.filter(candidateVersion => this._shouldPerformMigration(candidateVersion, previousMigratedVersion, versionToMigrate))
+			.sort(semver.compare);
+
+		newerVersions.forEach(version => {
+			try {
+				const migration = migrations[version];
+				migration(this);
+
+				this.set(MIGRATION_KEY, version);
```

Comment:
About the current implementation and idem-potency: after reading how the store works, it's actually atomic per change!

This means that we can restore the state of the store after a migration error. Something like this:

```js
const storeBackup = Object.assign({}, this.store);  // Or use spread if we can depending on the nodejs version we support
    
try {
    const store = Object.assign({}, this.store);  // Or use spread if we can depending on the nodejs version we support
    
    const migration = migrations[version];
    migration(this);

    this.set(MIGRATION_KEY, version);
} catch (err) {
    this.store = storeBackup;
    throw err;
}
```

**Still**, if `this.store = storeBackup;` fails for some reason (e.g.: disk full), the handler will run more than once. @sindresorhus can you weight your opinion here?

Reply:
It should only run once. Double-migrations can cause unforeseen problems.

Reply:
Alright @sindresorhus, will update it 😄